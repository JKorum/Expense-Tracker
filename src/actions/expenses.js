import uuid from 'uuid'

// ACTION GENERATORS 
// action gen --> action object --> arg for store.dispatch

const addExpense= (
	{
		description = ``, 
		note = ``, 
		amount = 0, 
		createdAt = 0
	} = {}
) => ({
	type: `ADD_EXPENSE`,
	expense: {
		id: uuid(),
		description,
		note,
		amount,
		createdAt
	}
})

const deleteExpense = (expenseId) => ({
	type: `DELETE_EXPENSE`,
	expenseId
})

const editExpense = (id, updates) => ({
	type: `EDIT_EXPENSE`, 
	id,
	updates
})

export { addExpense, deleteExpense, editExpense }