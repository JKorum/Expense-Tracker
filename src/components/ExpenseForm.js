import React from 'react'
import moment from 'moment'
import uuid from 'uuid'

import { SingleDatePicker } from 'react-dates'

class ExpenseForm extends React.Component {
	constructor(props) {
		super(props)
		this.state = {
			description: props.expense? props.expense.description : ``,
			note: props.expense? props.expense.note : ``,
			amount: props.expense? props.expense.amount.toString() : ``,
			createdAt: props.expense? moment(props.expense.createdAt) : moment(),
			calendarFocused: false,
			error: undefined,
			button: props.buttonName
		}

		this.handleDescriptionChange = this.handleDescriptionChange.bind(this)
		this.handleNoteChange = this.handleNoteChange.bind(this)
		this.handleAmountChange = this.handleAmountChange.bind(this)
		this.handleDateChange = this.handleDateChange.bind(this)
		this.handleFocusChange = this.handleFocusChange.bind(this)
		this.handleSubmit = this.handleSubmit.bind(this)
	}

	handleDescriptionChange(e) {
		const description = e.target.value
		this.setState(() => ({ description }))
	}

	handleNoteChange(e) {
		const note = e.target.value
		this.setState(() => ({ note }))
	}

	handleAmountChange(e) {
		const amount = e.target.value
		if (!amount || /^\d+(\.\d{0,2})?$/.test(amount)) { 
			this.setState(() => ({ amount }))
		}		
	}

	handleDateChange(calendarDate) {
		if (calendarDate) {
			this.setState(() => ({ createdAt: calendarDate }))
		}		
	}

	handleFocusChange({ focused }) {
		this.setState(() => ({ calendarFocused: focused }))
	}

	handleSubmit(e) {
		e.preventDefault()

		if (!this.state.description || !this.state.amount) {
			this.setState(() => ({ error: `Please provide description and amount` }))
		} else {
				this.setState(() => ({ error: undefined }))
				this.props.onSubmit({
					description: this.state.description,
					amount: +this.state.amount,
					createdAt: this.state.createdAt.valueOf(), // returns timestamp from moment obj
					note: this.state.note
				})
		}
	}

	render() {		
		return (
			<div>
				{this.state.error && <p>{this.state.error}</p>}
				<form onSubmit={this.handleSubmit}>
					<input 
						type="text" 
						placeholder="description" 
						autoFocus 
						value={this.state.description}
						onChange={this.handleDescriptionChange}
					/>

					<input 
						type="text" 
						value={this.state.amount}
						onChange={this.handleAmountChange}
						placeholder="amount" 
					/>

					<SingleDatePicker
					  date={this.state.createdAt} 
					  onDateChange={this.handleDateChange}
					  focused={this.state.calendarFocused}
					  onFocusChange={this.handleFocusChange}
					  id={uuid()}
					  numberOfMonths={1}		
					  isOutsideRange={() => false} // make days in a past clickable			  
					/>

					<textarea 
						value={this.state.note}
						onChange={this.handleNoteChange}
						placeholder="add a note for your expense"
					>
					</textarea>
					<button 
						type="submit"
						disabled={this.props.buttonName === `edit expense` && this.props.expense === undefined}
					>{this.state.button}</button>
				</form>
			</div>
		)
	}
}

export default ExpenseForm