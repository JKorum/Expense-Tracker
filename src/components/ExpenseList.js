import React from 'react'
import { connect } from 'react-redux'

import getVisibleExpenses from '../selectors/expenses'
import ExpenseListItem from './ExpenseListItem'

//component declaration props expects to be connected to store
export const ExpenseList = ({ expenses }) => (
	<div>
		{
			expenses.length === 0? (
				<p>No expenses</p>
			) : (
				expenses.map(expense => <ExpenseListItem key={expense.id} {...expense} />)
			) 
		}		
	</div>
)

// what store properties connected component should have access to
const mapStateToProps = (state) => ({
	expenses: getVisibleExpenses(state.expenses, state.filters)
})

//connect()() returns connected component 
export default connect(mapStateToProps)(ExpenseList)