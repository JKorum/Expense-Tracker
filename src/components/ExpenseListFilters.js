import React from 'react'
import { connect } from 'react-redux'
import uuid from 'uuid'
import { 
	setTextFilter, 
	sortByAmount, 
	sortByDate,
	setStartDate, 
	setEndDate 
} from '../actions/filters'

import { DateRangePicker } from 'react-dates'

//component declaration props expects to be connected to store
export class ExpenseListFilters extends React.Component {
	constructor(props) {
		super(props)
		this.state = {
			calendarFocused: null
		}

		this.handleCalendarDatesChange = this.handleCalendarDatesChange.bind(this)
		this.handleCalendarFocusChange = this.handleCalendarFocusChange.bind(this)
		this.handleTextChange = this.handleTextChange.bind(this)
		this.handleSortChange = this.handleSortChange.bind(this)
	}

	handleCalendarDatesChange({ startDate, endDate }) {		
		this.props.setStartDate(startDate)
		this.props.setEndDate(endDate)		
	}

	handleCalendarFocusChange(calendarFocused) {
		this.setState(() => ({ calendarFocused }))
	}

	handleTextChange(e) {		
		this.props.setTextFilter(e.target.value)
	}

	handleSortChange(e) {
		e.target.value === `date`? this.props.sortByDate() : this.props.sortByAmount() 
	}

	render() {
		const { filters, dispatch } = this.props

		return (
			<div>
				<input type="text" value={filters.text} onChange={this.handleTextChange} />
				<select value={filters.sortBy} onChange={this.handleSortChange}>
					<option value="date">Date</option>
					<option value="amount">Amount</option>
				</select>
				<DateRangePicker
				  startDate={filters.startDate} 
				  startDateId={uuid()} 
				  endDate={filters.endDate} 
				  endDateId={uuid()} 
				  onDatesChange={this.handleCalendarDatesChange}
				  focusedInput={this.state.calendarFocused}
				  onFocusChange={this.handleCalendarFocusChange}
				 	numberOfMonths={1}		
					isOutsideRange={() => false} // make days in a past clickable	
					showClearDates={true}		
				/>
			</div>
		)
	}
}

// what store properties connected component should have access to
const mapStateToProps = (state) => ({
	filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
	setStartDate: (startDate) => dispatch(setStartDate(startDate)),
	setEndDate: (endDate) => dispatch(setEndDate(endDate)),
	setTextFilter: (text) => dispatch(setTextFilter(text)),
	sortByDate: () => dispatch(sortByDate()),
	sortByAmount: () => dispatch(sortByAmount())
})

//connect()() returns connected component 
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)
