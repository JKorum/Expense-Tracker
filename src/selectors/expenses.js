import moment from 'moment'

// SELECTOR
// expenses (state), filters (state) --> selector --> subset of expenses based on filters 

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {	
	return expenses.filter(expense => {
		const createdAtMoment = moment(expense.createdAt) //timestamp --> moment obj 

		// if startDate === `undefined` --> startDateMatch === true --> will not effect result	
		const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, `day`) : true
		const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, `day`) : true
		const textMatch = text === `` || new RegExp(text, `i`).test(expense.description)

		return startDateMatch && endDateMatch && textMatch

	}).sort((a, b) => {
			if (sortBy === `date`) {
				return a.createdAt < b.createdAt ? 1 : -1 
			} else if (sortBy === `amount`) {
				return a.amount < b.amount ? 1 : (a.amount > b.amount? -1 : 0)
			}
	})	
}

export default getVisibleExpenses