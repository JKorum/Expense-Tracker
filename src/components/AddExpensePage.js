import React from 'react'
import { connect } from 'react-redux'

import ExpenseForm from './ExpenseForm'
import { addExpense } from '../actions/expenses'

export class AddExpensePage extends React.Component {
	constructor(props) {
		super(props)
		this.handleSubmit = this.handleSubmit.bind(this)
	}

	handleSubmit(expense) {
		this.props.addExpense(expense)
		this.props.history.push(`/`)
	}

	render() {
		return (
			<div>
				<h1>Add Expense</h1>
				<ExpenseForm 
					buttonName="add expense"
					onSubmit={this.handleSubmit} 
				/>
			</div>
		)
	}
}

const mapDispatchToProps = (dispatch) => ({
	addExpense: (expense) => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage)