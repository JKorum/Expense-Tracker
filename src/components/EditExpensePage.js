import React from 'react'
import { connect } from 'react-redux'

import ExpenseForm from './ExpenseForm'
import { editExpense, deleteExpense } from '../actions/expenses'

export class EditExpensePage extends React.Component {
	constructor(props) {
		super(props)
		this.handleSubmit = this.handleSubmit.bind(this)
		this.handleRemoveExpense = this.handleRemoveExpense.bind(this)
	}

	handleSubmit(expense) {		
		this.props.editExpense(this.props.expense.id, expense)
		this.props.history.push(`/`)
	}

	handleRemoveExpense(e){		
		this.props.deleteExpense(this.props.expense.id)
		this.props.history.push(`/`)
	}

	render() {
		return (
			<div>
				<h1>Edit Expense</h1>
				<ExpenseForm 
					buttonName="edit expense"
					expense={this.props.expense}
					onSubmit={this.handleSubmit}
				/>
				<button 
					type="button" 
					onClick={this.handleRemoveExpense} 
					disabled={!this.props.expense}
				>Remove</button>
			</div>
		)
	}
}

const mapStateToProps = (state, ownProps) => ({
	expense: state.expenses.find(expense => expense.id === ownProps.match.params.id)
})

const mapDispatchToProps = (dispatch) => ({
	editExpense: (id, expense) => dispatch(editExpense(id, expense)),
	deleteExpense: id => dispatch(deleteExpense(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)