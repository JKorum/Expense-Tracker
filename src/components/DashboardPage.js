import React from 'react'
import { connect } from 'react-redux'
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ExpensesSummary from './ExpensesSummary'
import getVisibleExpenses from '../selectors/expenses'

export const DashboardPage = ({ counter }) => (
	<div>
		{ counter > 0 && <ExpensesSummary /> }
		<ExpenseListFilters />
		<ExpenseList />
	</div>
)

const mapStateToProps = ({ expenses, filters }) => ({
	counter: getVisibleExpenses(expenses, filters).length
})

export default connect(mapStateToProps)(DashboardPage)