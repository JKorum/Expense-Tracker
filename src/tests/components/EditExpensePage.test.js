import React from 'react'
import { shallow } from 'enzyme'
import { EditExpensePage } from '../../components/EditExpensePage'
import expenses from '../fixtures/expenses'

let editExpenseSpy, deleteExpenseSpy, history, expense, wrapper

beforeEach(() => {
	editExpenseSpy = jest.fn()
	deleteExpenseSpy = jest.fn()
	history = { push: jest.fn() }
	expense = expenses[0]	

	const jsx = (
		<EditExpensePage 
			history={history} 
			expense={expense} 
			startEditExpense={editExpenseSpy} 
			startDeleteExpense={deleteExpenseSpy} 
		/>
	)		

	wrapper = shallow(jsx)
})

test(`should render EditExpensePage correctly`, () => {	
	expect(wrapper).toMatchSnapshot()
})

test(`should handle button click correctly`, () => {
	wrapper.find(`button`).simulate(`click`)
	expect(deleteExpenseSpy).toHaveBeenLastCalledWith(expense.id)
	expect(history.push).toHaveBeenLastCalledWith(`/`)
})

test(`should handle onSubmit correctly`, () => {
	wrapper.find(`ExpenseForm`).prop(`onSubmit`)(expense)
	expect(editExpenseSpy).toHaveBeenLastCalledWith(expense.id, expense)
	expect(history.push).toHaveBeenLastCalledWith(`/`)
})