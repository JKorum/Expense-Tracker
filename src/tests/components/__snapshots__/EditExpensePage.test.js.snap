// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <h1>
    Edit Expense
  </h1>
  <ExpenseForm
    buttonName="edit expense"
    expense={
      Object {
        "amount": "100",
        "createdAt": 0,
        "description": "hair tax",
        "id": "1",
        "note": "looks great",
      }
    }
    onSubmit={[Function]}
  />
  <button
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Remove
  </button>
</div>
`;
